File: 300.longest-increasing-subsequence.cpp
/*
 * @lc app=leetcode id=300 lang=cpp
 *
 * [300] Longest Increasing Subsequence
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class SegementTree {
public:
    int n;
    vector<int> tree;

    SegementTree(vector<int>& nums, int value, int L, int R) {
        n = 4 * nums.size();
        tree = vector<int>(n, 0);
    }

    void build(vector<int>& nums, int value, int L, int R) {
        if (L == R) {
            tree[value] = nums[L];
        } else {
            int M = (L + R) / 2;
            build(nums, value, L, M);
            build(nums, value, M+1, R);
            // How would you calculate the length of the longest subsequence?
        }
    }
};

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int globalMax = 0;

    }
};
// @lc code=end

File: 300.longest-increasing-subsequence.py
#
# @lc app=leetcode id=300 lang=python3
#
# [300] Longest Increasing Subsequence
#
from typing import List

# @lc code=start
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * (n)
        dp[0] = 1
        lenLIS = 1
        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j]+1)
            lenLIS = max(dp[i], lenLIS)
        return lenLIS

# @lc code=end

def main():
    s = Solution()
    s.lengthOfLIS([0,1,0,3,2,3])

if __name__ == '__main__':
    main()